### PR TITLE
Fix installation token cache issue

### PR DESCRIPTION
**Pull request by matthewgapp**
My application holds an Installation for an extended period of time (longer than 1 hour). This causes a "Bad credentials" error from Github. This PR introduces a fix that changes the installation token cache to only return a token if it's not expired, forcing Octocrab client to refresh the token appropriately.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -760,18 +761,54 @@ impl DefaultOctocrabBuilderConfig {
...
+            } else {
+                return Some(token.secret.clone());
+            }
+        }
+
+        None
+    }
+
+    fn valid_token(&self) -> Option<SecretString> {
+        self.valid_token_with_buffer(chrono::Duration::seconds(30))
```
Comment (matthewgapp) [inline src/lib.rs:806]: Consider a token that expires within 30 seconds to be expired by default. 

From what I can tell, the expiration is usually set about 1 hour from token mint.

File: src/lib.rs [hunk 2]
```diff
@@ -1349,7 +1386,21 @@ impl Octocrab {
 
         let token_object =
             InstallationToken::from_response(crate::map_github_error(response).await?).await?;
-        token.set(token_object.token.clone());
+
+        let expiration = token_object
+            .expires_at
+            .map(|time| {
+                DateTime::<Utc>::from_str(&time).map_err(|e| error::Error::Other {
```
Comment (matthewgapp) [inline src/lib.rs:1393]: Github doesn't document the timezone but I assumed it was UTC. Also checked a couple responses manually and verified that it was UTC.

**Comment by XAMPPRocky** [comment]
Thank you for your PR! However I need to check with someone to make sure these aren't overlapping (and different) proposals for the same feature.

cc @marcusirgens Is this problem solved by #436 or is it separate?

**Comment by marcusirgens** [comment]
@XAMPPRocky Yes, I solve this very similarily in #436, making `CachedToken::get` return `None` if the token is expired.

I'm not opposed to merging this first and working it into my PR, as it's more focused than my "do many things at once"-request, especially considering that #436 needs some clarifications. :)

https://github.com/marcusirgens/octocrab/blob/store-installation-tokens/src/lib.rs#L811-L828

**Comment by matthewgapp** [comment]
> I'm not opposed to merging this first and working it into my PR, as it's more focused than my "do many things at once"-request, especially considering that #436 needs some clarifications. :)

@XAMPPRocky sounds like it makes sense to merge so that @marcusirgens can rebase onto this.

**Comment by XAMPPRocky** [comment]
Thank you for your PR!

**Pull request merged**